### PR TITLE
Support different data precision between `UM_IC/PAR_IC` and the simulation field data

### PR DESCRIPTION
**Pull request by koarakawaii**
PR based on #264 
1. Add `2` runtime parameters `OPT__UM_IC_FLOAT8` and `PAR_IC_FLOAT8`
2. Both parameters are recorded in `HDF5` data dump and `Record__Note`
3. Use `load_data_size` to determine data size when access `float/double` IC data; use `char *` pointer to when accessing `float/double` IC data to maintain flexibility
4. Default behavior when `OPT__UM_IC_FLOAT8` and `PAR_IC_FLOAT8` are not set is to set them equal to `FLOAT8`

### CODE REVIEW

**Review by JeiwHuang — APPROVED** [review]
I have done the tests.

- I used `CDM_LSS` to test 4 combinations of `FLOAT8` and `PAR_IC_FLOAT8`, with the original and converted-to-double-precision `PAR_IC`.
   - The figures for the results all look normal.
   - The new option `PAR_IC_FLOAT8` can work as my expectation.
- I used `EnergyPowerSpectrum` to test the 4 combinations of `FLOAT8` and `OPT__UM_IC_FLOAT8`, with the single or double `UM_IC` I created.
   - The power spectra of the results are correct, with only machine precision differences.
   - The new option `OPT__UM_IC_FLOAT8` can work as my expectation.
   
I did not find issues with this PR.

File: src/Particle/Par_Init_ByFile.cpp [hunk 1]
```diff
@@ -72,11 +72,14 @@ void Par_Init_ByFile()
    if ( !Aux_CheckFileExist(FileName) )
       Aux_Error( ERROR_INFO, "file \"%s\" does not exist for PAR_INIT == PAR_INIT_BY_FILE !!\n", FileName );
 
+// determine the load_data_size for loading PAR_IC
+   size_t load_data_size   = ( PAR_IC_FLOAT8 ) ? sizeof(double) : sizeof(float);
```
Comment (hyschive) [inline src/Particle/Par_Init_ByFile.cpp:76]: Single space between `load_data_size` and `=`

File: src/Particle/Par_Init_ByFile.cpp [hunk 2]
```diff
@@ -129,13 +133,30 @@ void Par_Init_ByFile()
 //    collect data for the target particle
 //    [id][att]
       if ( amr->Par->ParICFormat == PAR_IC_FORMAT_ID_ATT )
-         memcpy( ParData1, ParData_ThisRank+p*NParAtt, NParAtt*sizeof(real) );
+      {
+         if ( sizeof(real) == load_data_size )
+            memcpy( ParData1, ParData_ThisRank+p*NParAtt*load_data_size, NParAtt*load_data_size );
+         else
```
Comment (hyschive) [inline src/Particle/Par_Init_ByFile.cpp:139]: It seems redundant. Can we remove it and rely on `if ( PAR_IC_FLOAT8 ) ... else ...`?

File: src/Particle/Par_Init_ByFile.cpp [hunk 3]
```diff
@@ -129,13 +133,30 @@ void Par_Init_ByFile()
...
+      }
 
 //    [att][id]
       else
       {
-         for (int v=0; v<NParAtt; v++)
-            ParData1[v] = ParData_ThisRank[ v*NParThisRank + p ];
+         if ( sizeof(real) == load_data_size )
+            for (int v=0; v<NParAtt; v++) ParData1[v] = *((real*)(ParData_ThisRank+(v*NParThisRank+p)*load_data_size));
+         else
```
Comment (hyschive) [inline src/Particle/Par_Init_ByFile.cpp:153]: It seems redundant. Can we remove it and rely on `if ( PAR_IC_FLOAT8 ) ... else ...`?

File: src/Particle/Par_Init_ByFile.cpp [hunk 4]
```diff
@@ -97,21 +100,22 @@ void Par_Init_ByFile()
...
-   for (int r=0; r<MPI_Rank; r++)   FileOffset = FileOffset + long(NParAttPerLoad)*NPar_EachRank[r]*sizeof(real);
+   for (int r=0; r<MPI_Rank; r++)   FileOffset = FileOffset + long(NParAttPerLoad)*NPar_EachRank[r]*load_data_size;
 
 
 // load data
    if ( MPI_Rank == 0 )    Aux_Message( stdout, "   Loading data ... " );
 
-   real *ParData_ThisRank = new real [ NParThisRank*NParAtt ];
+// allocate buffer for loading PAR_IC
+   char* ParData_ThisRank = new char [NParThisRank*NParAtt*load_data_size];
```
Comment (hyschive) [inline src/Particle/Par_Init_ByFile.cpp:110]: Add a space after `[` and before `]`

File: src/Particle/Par_Init_ByFile.cpp [hunk 5]
```diff
@@ -97,21 +100,22 @@ void Par_Init_ByFile()
...
 
 // note that fread() may fail for large files if sizeof(size_t) == 4 instead of 8
    FILE *File = fopen( FileName, "rb" );
 
    for (int v=0; v<NParAtt; v+=NParAttPerLoad)
    {
-      fseek( File, FileOffset+v*NParAllRank*sizeof(real), SEEK_SET );
-      fread( ParData_ThisRank+v*NParThisRank, sizeof(real), long(NParAttPerLoad)*NParThisRank, File );
+      fseek( File, FileOffset+v*NParAllRank *load_data_size, SEEK_SET );
```
Comment (hyschive) [inline src/Particle/Par_Init_ByFile.cpp:117]: Remove the space before `*`

File: src/Particle/Par_Init_ByFile.cpp (hunk 3 above)
Comment (koarakawaii) [reply src/Particle/Par_Init_ByFile.cpp:153]: Yes, I think we can 😅 .

File: src/Particle/Par_Init_ByFile.cpp (hunk 2 above)
Comment (koarakawaii) [reply src/Particle/Par_Init_ByFile.cpp:139]: @hyschive : I think if there is no merit for using `memcpy` to copy the whole consecutive memory than copying elementwise by `for` loop when filling `ParData1`, we can remove this `memcpy` line.

File: src/Init/Init_ByFile.cpp [hunk 6]
```diff
@@ -569,15 +577,32 @@ void Init_ByFile_AssignData( const char UM_Filename[], const int UM_lv, const in
...
 
-                  Offset_PG = (long)NVarPerLoad*IDX321( i+Disp_i, j+Disp_j, k+Disp_k, PS2, PS2 );
+                  Offset_PG = (long)NVarPerLoad*IDX321( i+Disp_i, j+Disp_j, k+Disp_k, PS2, PS2 )*load_data_size;
 
                   if ( UM_Format == UM_IC_FORMAT_ZYXV )
-                     memcpy( fluid_in, PG_Data+Offset_PG, UM_NVar*sizeof(real) );
+                  {
+                     if ( sizeof(real) == load_data_size )
+                        memcpy( fluid_in, PG_Data+Offset_PG, UM_NVar*load_data_size );
+                     else
```
[the remark this answers is not in the document]
Comment (koarakawaii) [reply src/Init/Init_ByFile.cpp:586]: @hyschive : I think if there is no merit for using `memcpy` to copy the whole consecutive memory than copying elementwise by `for` loop when filling `fluid_in`, we can remove this `memcpy` line.

File: src/Init/Init_ByFile.cpp [hunk 7]
```diff
@@ -569,15 +577,32 @@ void Init_ByFile_AssignData( const char UM_Filename[], const int UM_lv, const in
...
+                     }
+                  }
 
                   else
                   {
-                     for (int v=0; v<UM_NVar; v++)
-                        fluid_in[v] = *( PG_Data + Offset_PG + v*CUBE(PS2) );
+                     if ( sizeof(real) == load_data_size )
+                        for (int v=0; v<UM_NVar; v++) fluid_in[v] = *((real*)( PG_Data + Offset_PG + v*CUBE(PS2)*load_data_size ));
+                     else
```
[the remark this answers is not in the document]
Comment (koarakawaii) [reply src/Init/Init_ByFile.cpp:599]: Yes, I think we can 😅 .

File: src/Init/Init_ByFile.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (hyschive) [reply src/Init/Init_ByFile.cpp:586]: Then let's remove it. Performance is not of concern during initialization.

File: src/Particle/Par_Init_ByFile.cpp (hunk 2 above)
Comment (hyschive) [reply src/Particle/Par_Init_ByFile.cpp:139]: Let's remove it

**Comment by koarakawaii** [comment]
@hyschive Modifications have been made according to the review comment (except for those `[Reminder]`s)~

**Review by hyschive — APPROVED** [review]
@koarakawaii Looks great. Thanks for the contribution!

**Pull request merged**